### ISSUE TITLE
Infinite loop of queueing swank commands when starting Slime under NetBSD (analysis + workaround only)

### ISSUE THREAD

**Issue by aymeric-nb** [slime/slime#600]
Hi,

under NetBSD, IPv4 and IPv6 live completely separately and their port numbers are also allocated in different pools.

The common lisp started by slime will listen on some port on "localhost" and tell slime the port number. Slime then connects to this port on "localhost". However, the "localhost" determined by swank (at least in the sbcl backend, didn't check others) might be IPv4 or IPv6, and the "localhost" determined by slime might be the other IP version. Usually, slime will try the other IP version if the first connect fails and everything will "just work".

But, when a bare NetBSD starts and no network activity has taken place, both IPv4 and IPv6 port numbers are synchronized, so swank will listen on an IP stack and slime (unfortunately) will connect to the other IP stack. The funny thing here is that slime (as a client) gets assigned the same port number as swank (but on the other stack) and so it starts talking to itself on its socket. Emacs' status bar shows an increasing number of swank requests being queued)

Here are the potential fixes I can think of:
 - make sure swank and slime use the same deterministic way of choosing an IP stack (fragile but easiest solution)
 - change the initial information from swank to slime to include the resolved IP address in addition to the port number, so slime knows precisely where to connect
 - maybe extend this information to include several potential (address . port) pairs and have swank listen on every possible "localhost"

Workaround: open any network socket (e.g. nc localhost 9999 with 9999 not necessarily open) on any of the two IP stacks. This will decorrelate the port numbers and slime will work perfectly. [ Swapping the IPv4 and IPv6 entries of localhost in /etc/hosts also works but is way more intrusive. ]

Under Linux, listening on localhost on an IPv6 socket implies listening also on the same port in the IPv4 stack so you won't be able to reproduce the problem there. Please read "Interaction between IPv4/v6 sockets" in http://man.netbsd.org/inet6.4 before brushing this off as a NetBSD bug.

The swank backends are per-implementation and I tried this only with SBCL under NetBSD, other combinations may be affected.

Regards,

**Comment by luismbo** [comment]
Your second fix suggestion sounds good. Listening on multiple interfaces doesn't seem very useful, users can pick an explicit interface if they want to force IPv6 or IPv4. Would you like to make a pull request?

**Comment by aymeric-nb** [comment]
Honestly, I think it's above my lisp skills to properly handle the parsing of different IP address families and updating the backends, so please move forward if you have a fix. I might come up with a pull request later but I'll be perfectly happy if it's solved by then.

**Comment by luismbo** [comment]
This would involve adding a new `definterface local-address` to swank/backend.lisp, implementing it for each backend (the usocket library should know how) and use it in `setup-server` (and places that provide an announce-fn such as `start-server` and `simple-announce-function`).

flip214 (on irc.freenode.net, I don't know his github handle) was looking into this for similar reasons. Perhaps he's started working on it.

**Comment by luismbo** [comment]
@aymeric-nb indeed, @phmarek has gotten started. https://github.com/phmarek/slime/commit/6c9634aa4975462d7be936daddbf3483c57c67eb It's missing the elisp side.

**Comment by aymeric-nb** [comment]
So I wrote some basic emacs lisp counterpart to support parsing a "address port" line, you can get it here:
aymeric-nb/slime@aa3247c58343826b0a65effed1dfb4c321a84a7b

It would be more elegant to use seq-let to bind address and port in two places, but this requires pulling seq.el so I preferred to keep the footprint of the commit as small as possible. Feel free to improve on it.

Regards

**Comment by aymeric-nb** [comment]
Also, there is a (very short?) race if the FORMAT or underlying write(s) on the SWANK server side are not atomic; IMHO a better way would be to create a slime.PID.tmp file and rename it to slime.PID once it's filled. This would also avoid checking the size of the file before reading it in slime-attempt-connection.
